rllib/optimizers/segment_tree.py

```diff
@@ -2,140 +2,149 @@
+        """Applies `self.operation` to subsequence of our values.
+
+        Subsequence is contiguous, includes `start` and excludes `end`.
 
           self.operation(
               arr[start], operation(arr[start+1], operation(... arr[end])))
 
-        Parameters
-        ----------
-        start: int
-          beginning of the subsequence
-        end: int
-          end of the subsequences
-
-        Returns
-        -------
-        reduced: obj
-          result of reducing self.operation over the specified range of array
-          elements.
+        Args:
+            start (int): Start index to apply reduction to.
+            end (int): Start index to apply reduction to (excluded).
+
+        Returns:
+            any: The result of reducing self.operation over the specified
+                range of `self._value` elements.
         """
         if end is None:
-            end = self._capacity - 1
-        if end < 0:
-            end += self._capacity
-        return self._reduce_helper(start, end, 1, 0, self._capacity - 1)
+            end = self.capacity
+        elif end < 0:
+            end += self.capacity
+
+        # Init result with neutral element.
+        result = self.neutral_element
+        start += self.capacity
+        end += self.capacity
+
+        while start < end:
+            if start & 1:
+                result = self.operation(result, self.value[start])
+                start += 1
+            if end & 1:
+                end -= 1
+                result = self.operation(result, self.value[end])
+            start = start >> 1
+            end = end >> 1
+        return result
 
     def __setitem__(self, idx, val):
-        # index of the leaf
-        idx += self._capacity
-        self._value[idx] = val
-        idx //= 2
+        """
+        Inserts/overwrites a value in/into the tree.
+
+        Args:
+            idx (int): The index to insert to. Must be in [0, `self.capacity`[
+            val (float): The value to insert.
+        """
+        assert 0 <= idx < self.capacity
+
+        # Index of the leaf to insert into (always insert in "second half"
+        # of the tree, the first half is reserved for already calculated
+        # reduction-values).
+        idx += self.capacity
+        self.value[idx] = val
+
+        # Reclculate all affected reduction values (in "first half" of tree).
+        idx = idx >> 1  # Divide by 2 (fater than division).
         while idx >= 1:
-            self._value[idx] = self._operation(self._value[2 * idx],
-                                               self._value[2 * idx + 1])
-            idx //= 2
+            update_idx = 2 * idx  # calculate only once
+            # Update the reduction value at the correct "first half" idx.
+            self.value[idx] = self.operation(self.value[update_idx],
+                                             self.value[update_idx + 1])
+            idx = idx >> 1  # Divide by 2 (fater than division).
```

Comment:
Given that this code is written in python, I kind of doubt using div is helping much. Does it make sense to revert to floordiv for code clarity?

Reply:
Indeed, in the microbenchmarks, it seems like `idx //= 2` is slightly faster than `idx = idx >> 1`, not sure why and whether it's really significant, but I changed it back to `//= 2` either way.
Btw, the new tree is about 2.7 times faster than the old one in summing up over some larger range with capacity=100k.